build: add a script for setting the release version in examples, charts,  and docs

obnoxxx opened the pull request:
**Description of changes:**  
  This is adds the first incarnation of  of a script to prepare examples and docs for a new release.

    invocation in principle: set-release-ver.sh NEW_VER
    concrete example: set-release-ver.sh v1.16.8

The originating version ios auto-detected.

**Issues resolved by this PR:** 

Resolves #15749 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env bash
```
This is the standard set of bash script options we should use for all scripts, except in special cases. These could be important for helping the person running the script identify when an issue has occurred.

```suggestion
#!/usr/bin/env bash
set -euo pipefail
```

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env bash
+
+# This script is supposed to facilitate the preparation of a rook release by
+# changing example and documentation files to use the new release version instead the old one.
+#
+# It will usually be run on a release branch and the resulting changes will be the content of the
+# commit for the basis of the release.
+#
+#
+
+
+
+
+
+# The script needs two version tags:
+
+# FROM_TAG: the version tag of the previous/current release (e. g. v1.16.7)
+#   TO_TAG: the version tag of the new release being prepared (e. g. v1.16.8)   
+#
+# The script  replaces all occurrences of $FROM_TAG with  $TO_TAG in
+# the relevant affected files.
+
+
+# The two tags can be passed to the script as the first two positional parameters in its invocation.
+
+# So an example invocation could be:
+# set-release-ver.sh v1.16.7 v1.16.8
```
Can this be made less verbose?

obnoxxx replied:
@BlaineEXE 
> Can this be made less verbose?

It is a lot less verbose now.

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,64 @@
+# set-release-ver.sh v1.16.7 v1.16.8
+
+
+FROM_TAG="$1"
+TO_TAG="$2"
```
Personally, I prefer to put args right under the shebang and `set` line (separated by an empty line) so that it acts as its own documentation. From these lines, I can interpret immediately that there are 2 args and what they are named for the rest of the script.

obnoxxx replied:
done

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,64 @@
+
+
+FILES="
+ Documentation/Getting-Started/quickstart.md
+ Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
+ Documentation/Upgrade/rook-upgrade.md
+ deploy/charts/rook-ceph/values.yaml
+ deploy/examples/direct-mount.yaml
+ deploy/examples/images.txt
+ deploy/examples/operator-openshift.yaml
+ deploy/examples/operator.yaml
+ deploy/examples/osd-purge.yaml
+ deploy/examples/toolbox-job.yaml
+ deploy/examples/toolbox-operator-image.yaml
```
What's with the single space before each line here?

obnoxxx replied:
removed.

BlaineEXE reviewed (changes requested):
I'm getting this output on macos

```
FROM_TAG: 'master'
  TO_TAG: 'v1.18.0-beta.0'
sed: 
 Documentation/Getting-Started/quickstart.md
 Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
 Documentation/Upgrade/rook-upgrade.md
 deploy/charts/rook-ceph/values.yaml
 deploy/examples/direct-mount.yaml
 deploy/examples/images.txt
 deploy/examples/operator-openshift.yaml
 deploy/examples/operator.yaml
 deploy/examples/osd-purge.yaml
 deploy/examples/toolbox-job.yaml
 deploy/examples/toolbox-operator-image.yaml
: No such file or directory
```

BlaineEXE commented:
When I fix the script to run correctly, this is making changes that we don't want to make using `master` and `v1.18.0-beta.0` as inputs. See example below to see an undesirable change (not made in [Travis's example](https://github.com/rook/rook/pull/15596/files) along with a desirable change.

<img width="1259" alt="image" src="https://github.com/user-attachments/assets/96d365cc-535a-407e-8d10-fd13069979e0" />

I am going to mark this PR as draft. Please revisit the implementation and test thoroughly to make sure it matches Travis's example. A naive sed replace `$1` with `$2` is a bit too broad. I suspect `vx.y.1`, and `vx.y.2` are going to be corner cases as well in the upgrade doc. Some examples of those are here: 
https://github.com/rook/rook/pull/15229/files
https://github.com/rook/rook/pull/15290/files

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,45 @@
+
+
+
+FILES="
+Documentation/Getting-Started/quickstart.md
+Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
+Documentation/Upgrade/rook-upgrade.md
+deploy/charts/rook-ceph/values.yaml
+deploy/examples/direct-mount.yaml
+deploy/examples/images.txt
+deploy/examples/operator-openshift.yaml
+deploy/examples/operator.yaml
+deploy/examples/osd-purge.yaml
+deploy/examples/toolbox-job.yaml
+deploy/examples/toolbox-operator-image.yaml
+"
+
+sed -i -e "s/$FROM_TAG/$TO_TAG/g" $FILES
```
Using an array resolves the execution error I get on macos.

```suggestion
FILES=(
Documentation/Getting-Started/quickstart.md
Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
Documentation/Upgrade/rook-upgrade.md
deploy/charts/rook-ceph/values.yaml
deploy/examples/direct-mount.yaml
deploy/examples/images.txt
deploy/examples/operator-openshift.yaml
deploy/examples/operator.yaml
deploy/examples/osd-purge.yaml
deploy/examples/toolbox-job.yaml
deploy/examples/toolbox-operator-image.yaml
)

sed -i -e "s/$FROM_TAG/$TO_TAG/g" "${FILES[@]}"
```

obnoxxx commented:
The `No such file or directory`  error from sed is due to the double quoting that I added around `$FILES` that I added to silence shellcheck.  But I think in this case the word splitting that the quoting prevents  is exactly what the sed invocation needs to function correctly

obnoxxx commented:
@BlaineEXE  wrote:
> > The No such file or directory error from sed is due to the double quoting that I added around $FILES that I added to silence shellcheck. But I think in this case the word splitting that the quoting prevents is exactly what the sed invocation needs to function correctly
> 
> Using an array and using it as `"${FILES[@]}"` resolves the quoting issue while still ensuring any potential paths with spaces will be quoted as sed arguments.

Applied. Thanks!

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,45 @@
+Documentation/Getting-Started/quickstart.md
+Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
+Documentation/Upgrade/rook-upgrade.md
+deploy/charts/rook-ceph/values.yaml
+deploy/examples/direct-mount.yaml
+deploy/examples/images.txt
+deploy/examples/operator-openshift.yaml
+deploy/examples/operator.yaml
+deploy/examples/osd-purge.yaml
+deploy/examples/toolbox-job.yaml
+deploy/examples/toolbox-operator-image.yaml
```
Probably a future update, but let's consider `deploy/examples/*.yaml` instead of individualized yamls. It's much more frequent that we add yamls that might have images.

The benefit of the yamls is that images should always have `image:` preceding them, and they should be in some format like `rook/ceph:tag`, which can be used to prevent false positives.

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+FROM_TAG="$1"
```
I'm not sure this is so simple when we cut a `v1.Y.0-beta.0` tag from `master` when it's already been branched to `release-v1.Y`. It seems like it might be good to add autodetection later and focus first on getting the find/replace logic correct.

obnoxxx commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,45 @@
+deploy/examples/toolbox-operator-image.yaml
+)
+
+sed -i -e "s/$FROM_TAG/$TO_TAG/g" "${FILES[@]}"
```
@travisn  wrote:
> This command will be too general. If the tag is `master`, we will end up replacing too many references. Instead, let's use very specific `sed` commands to avoid changing too many places. For example:
> 
> ```
> "s/docker.io/rook/ceph:$FROM_TAG/docker.io/rook/ceph:$TO_TAG/g" 
> "s/tag: $FROM_TAG/tag: $TO_TAG/g" 
> ```

sounds good. I was trying to implement something along those lines.

But I think this might be too narrow: It might miss the documentation pieces.

I might end up using different sed rules for different sets of files. - wdyt?


> And there are others, basically we should end up with the same results as we see in a previous PR such as #15737. It should be simple to run the script and inspect that we get the same results as that PR.

obnoxxx commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,49 @@
+deploy/charts/rook-ceph/values.yaml
+)
+FILES_EXAMPLES=(
+deploy/examples/direct-mount.yaml
```
will try and report back

obnoxxx replied:
tested and it looks good to me. PR updated accordingly.

obnoxxx replied:
In fact I replaced it with a command `finddeploy/examples -type f`now.

obnoxxx commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,49 @@
+# It will usually be run on a release branch and the resulting changes will be the content of the
+# commit for the basis of the release.
+
+# The script needs two version tags:
```
iu ses two but only one needs to be passed as argument. Wording updated.

obnoxxx commented:
@BlaineEXE  wrote:
> When I fix the script to run correctly, this is making changes that we don't want to make using `master` and `v1.18.0-beta.0` as inputs. See example below to see an undesirable change (not made in [Travis's example](https://github.com/rook/rook/pull/15596/files) along with a desirable change.
> <img alt="image" width="1259" src="https://private-user-images.githubusercontent.com/3989526/437549903-96d365cc-535a-407e-8d10-fd13069979e0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU5NTAyMzksIm5iZiI6MTc0NTk0OTkzOSwicGF0aCI6Ii8zOTg5NTI2LzQzNzU0OTkwMy05NmQzNjVjYy01MzVhLTQwN2UtOGQxMC1mZDEzMDY5OTc5ZTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MjlUMTgwNTM5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzY3ODMzMTRkNzlmOGE1ZjAzYmE2ZGE1Nzc5ZTIyNDQ1ZDA0MDEzMWU2N2E4MGRhNzk2ZjE3ODBjYmY1YzJjNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.qKB1lTyAl8zpP00cWB1RUgw1EppBcEMbhab0zdUTDQs">
> 
> I am going to mark this PR as draft. Please revisit the implementation and test thoroughly to make sure it matches Travis's example. A naive sed replace `$1` with `$2` is a bit too broad. I suspect `vx.y.1`, and `vx.y.2` are going to be corner cases as well in the upgrade doc. Some examples of those are here: https://github.com/rook/rook/pull/15229/files https://github.com/rook/rook/pull/15290/files

Thanks for pointing this out!

It took me a while to understan the problem here. 

The latest version of the script leaves those liks alone.

obnoxxx commented:
I think  the broader requests and suggestions have been addressed.

The script seems to be working correctly in my local tests now as far as I can tell.

removing Draft ...

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,43 @@
+FROM_TAG=$(grep "docker.io/rook/ceph" deploy/examples/images.txt | awk -F : '{ print $2 }')
+TO_TAG="$1"
+
+# This script is supposed to facilitate the preparation of a rook release by
+# changing example and documentation files to use the new release version instead the old one.
+#
+# It will usually be run on a release branch and the resulting changes will be the content of the
+# commit for the basis of the release.
+
+# The script uses two version tags:
+
+# FROM_TAG: the version tag of the previous/current release (e. g. v1.16.7)
+# TO_TAG: the version tag of the new release being prepared (e. g. v1.16.8)
+
+# TO_TAG should be passed to the script as the first positional parameter in its invocation.
+
+# So an example invocation could be:
+# set-release-ver.sh v1.17.2
+#
+# This assumes invocation from the release-1.17 branch with latest
+# release tag v1.17.1
+#
+# The script replaces all occurrences of $FROM_TAG with $TO_TAG in
+# the relevant affected files.
```
Can this be summarized in fewer lines? 

I can see several things that can be cleaned up:
1. FROM_TAG is no longer an input and thus doensn't need documentation
2. invocation isn't needed because it's documented in the release  readme
3. don't need to state TO_TAG is final positional because that's implicitly documented above

In addition, just some repetition that can be reduced. I think this could be 2-3 lines and still perfectly helpful

obnoxxx replied:
done.

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+FROM_TAG=$(grep "docker.io/rook/ceph" deploy/examples/images.txt | awk -F : '{ print $2 }')
+TO_TAG="$1"
```
Putting input args at the top serves to make them more self documenting. Because FROM_TAG is no longer an input arg (but still critical to understanding the function), I would suggest making it the second section. A comment also wouldn't hurt

```suggestion
TO_TAG="$1"

FROM_TAG=$(grep "docker.io/rook/ceph" deploy/examples/images.txt | awk -F : '{ print $2 }')

```

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,43 @@
+# The script replaces all occurrences of $FROM_TAG with $TO_TAG in
+# the relevant affected files.
+
+FILES_DOCS=(
+Documentation/Getting-Started/quickstart.md
+Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
+Documentation/Upgrade/rook-upgrade.md
+)
+FILES_CHARTS=(
+deploy/charts/rook-ceph/values.yaml
+)
+FILES_EXAMPLES=(
+    deploy/examples/*
+)
```
Please pick an indentation and stick with it. Personally, indented looks best to me. I'd also suggest make sure the indentation is 2 spaces, to match other rook bash scripts like `github-action-helper.sh`.

```suggestion
FILES_DOCS=(
  Documentation/Getting-Started/quickstart.md
  Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
  Documentation/Upgrade/rook-upgrade.md
)
FILES_CHARTS=(
  deploy/charts/rook-ceph/values.yaml
)
FILES_EXAMPLES=(
  deploy/examples/*
)
```

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,43 @@
+    deploy/examples/*
+)
+
+sed -i -e "s/docker.io\/rook\/ceph:$FROM_TAG/docker.io\/rook\/ceph:$TO_TAG/g" "${FILES_EXAMPLES[@]}"
+sed -i -e  "s/tag: $FROM_TAG/tag: $TO_TAG/g" "${FILES_CHARTS[@]}"
+sed -i -e  "s/ $FROM_TAG / $TO_TAG /g" "${FILES_DOCS[@]}"
```
Nit: variables are declared in opposite order that sed commands are called. I'd suggest ordering them the same to help with legibility

obnoxxx replied:
gppd point. changed.

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,43 @@
+    deploy/examples/*
+)
+
+sed -i -e "s/docker.io\/rook\/ceph:$FROM_TAG/docker.io\/rook\/ceph:$TO_TAG/g" "${FILES_EXAMPLES[@]}"
+sed -i -e  "s/tag: $FROM_TAG/tag: $TO_TAG/g" "${FILES_CHARTS[@]}"
+sed -i -e  "s/ $FROM_TAG / $TO_TAG /g" "${FILES_DOCS[@]}"
```
Please continue to be vigilant for unnecessary double spaces

```suggestion
sed -i -e "s/docker.io\/rook\/ceph:$FROM_TAG/docker.io\/rook\/ceph:$TO_TAG/g" "${FILES_EXAMPLES[@]}"
sed -i -e "s/tag: $FROM_TAG/tag: $TO_TAG/g" "${FILES_CHARTS[@]}"
sed -i -e "s/ $FROM_TAG / $TO_TAG /g" "${FILES_DOCS[@]}"
```

BlaineEXE commented on build/release/README.md:
```diff
@@ -79,6 +79,12 @@ After the PR is merged, you can tag the release with the beta tag (`v1.13.0-beta
 
 **IMPORTANT** Before tagging the release, open a new PR to update the documentation and example manifest tags to the release version.
 
+The script  `set-release-ver.sh` can help preparing these changes:
```
Have you looked into enabling whitespace visibility in your editor yet to help prevent double space issues? Please do this.

```suggestion
The script `set-release-ver.sh` can help preparing these changes:
```

obnoxxx replied:
@BlaineEXE wrote:
> Have you looked into enabling whitespace visibility in your editor yet to help prevent double space issues?

yes but no conclusion yet.


> Please do this.

I will continue and find a solution

BlaineEXE commented on build/release/README.md:
````diff
@@ -79,6 +79,12 @@ After the PR is merged, you can tag the release with the beta tag (`v1.13.0-beta
 
 **IMPORTANT** Before tagging the release, open a new PR to update the documentation and example manifest tags to the release version.
 
+The script  `set-release-ver.sh` can help preparing these changes:
+
+```console
+# Pass the new version
````
I think "pass the new version" is better left as an instruction above. It'll be easily copy-pasted with the line below as part of the snippet. Seems best to leave it so it won't risk causing any issues when that happens

obnoxxx replied:
done

BlaineEXE commented on build/release/README.md:
````diff
@@ -79,6 +79,12 @@ After the PR is merged, you can tag the release with the beta tag (`v1.13.0-beta
+
+```console
+# Pass the new version
+$ tests/scripts/set-release-ver.sh v1.16.8
````
Other console snippets above don't use dollar signs. If we leave this out copying and pasting is easier with less friction

```suggestion
tests/scripts/set-release-ver.sh v1.16.8
```

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env bash
```
I was wondering when/if this would come up. Mac's sed version is POSIX/BSD sed, not gnu-sed, and the -i option is different in gnu.

This may need the `-i ''` option as noted here: https://www.baeldung.com/linux/gnu-bsd-stream-editor#2-the--i-option

obnoxxx replied:
I tried it but the main problem currently seems to be that sed refuses top operate on non-regular filel like directory  `deploy/examples/cosi` which are expanded by `deploy/examples/*`.

I will figure it uot and update.

obnoxxx replied:
FWIW: works just fine for me on Linux where I developed it initially. will take care of MacOS next.

obnoxxx replied:
MacOS works again now.

obnoxxx replied:
@BlaineEXE  wrote:
> I was wondering when/if this would come up. Mac's sed version is POSIX/BSD sed, not gnu-sed, and the -i option is different in gnu.
> 
> This may need the `-i ''` option as noted here: https://www.baeldung.com/linux/gnu-bsd-stream-editor#2-the--i-option

after struggling with a correct invocation  for a while, I ended up uysing the `build/sed-in-place` script.

obnoxxx commented on build/release/README.md:
````diff
@@ -79,6 +79,12 @@ After the PR is merged, you can tag the release with the beta tag (`v1.13.0-beta
+```console
+# Pass the new version
+$ tests/scripts/set-release-ver.sh v1.16.8
+```
````
done

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,28 @@
+
+$SED "s/ $FROM_TAG / $TO_TAG /g" "${FILES_DOCS[@]}"
+$SED "s/tag: $FROM_TAG/tag: $TO_TAG/g" "${FILES_CHARTS[@]}"
+$SED "s/docker.io\/rook\/ceph:$FROM_TAG/docker.io\/rook\/ceph:$TO_TAG/g" "${FILES_EXAMPLES[@]}"
```
Alternately, I think this could be one line which could avoid duplicating the search-replace expression

```
$SED "s/docker.io\/rook\/ceph:$FROM_TAG/docker.io\/rook\/ceph:$TO_TAG/g" deploy/examples/*.yaml deploy/examples/images.txt
```

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,32 @@
+
+FROM_TAG=$(grep "docker.io/rook/ceph" deploy/examples/images.txt | awk -F : '{ print $2 }')
+
+scriptdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+SED="${scriptdir}/../../build/sed-in-place"
```
This raises a good point. `SED` is defined with an absolute path, which is good form. I can foresee someone running this script from the `tests/scripts` dir and then it not working correctly. (aside: @travisn should this reside next to `sed-in-place` in the build dir instead? IMO yes.)

Given that, I would suggest a different strategy that will ensure the whole script works, not just the sed portion.

```suggestion
scriptdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
rook_root="${scriptdir}/../..}"
cd "${rook_root}" # allow using relative paths in script

SED="build/sed-in-place"
```

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,32 @@
+ deploy/examples/images.txt
+)
+
+$SED -e "s/rook\/ceph:$FROM_TAG/rook\/ceph:$TO_TAG/g" -e "s/rook-version=$FROM_TAG/rook-version=$TO_TAG/g" Documentation/Upgrade/rook-upgrade.md
```
It's easier to read separate sed expressions when they're broken up to one expression per line

```suggestion
$SED -e "s/rook\/ceph:$FROM_TAG/rook\/ceph:$TO_TAG/g" \
     -e "s/rook-version=$FROM_TAG/rook-version=$TO_TAG/g" \
  Documentation/Upgrade/rook-upgrade.md
```

BlaineEXE reviewed (changes requested):
shellcheck is failing

BlaineEXE commented on tests/scripts/set-release-ver.sh:
```diff
@@ -0,0 +1,54 @@
+$SED "s/tag: $FROM_TAG/tag: $TO_TAG/g" "${FILES_CHARTS[@]}"
+$SED "s/docker.io\/rook\/ceph:$FROM_TAG/docker.io\/rook\/ceph:$TO_TAG/g" "${FILES_EXAMPLES[@]}"
+
+# Special processing for upgrade documentation
+
+# Use a different FROM_TAG for some of the upgrade documentation updates
+# Extract it from a line that shows the version in this form (with the backticks):
+#  `rook-version=v1.17.0`
+FROM_TAG_UPGRADE=$(grep "\`rook-version=" Documentation/Upgrade/rook-upgrade.md | awk -F = '{ print $2 }' | awk -F "\`" '{ print $1 }')
+
+$SED -e "s/rook\/ceph:$FROM_TAG/rook\/ceph:$TO_TAG/g" \
+     -e "s/rook-version=$FROM_TAG/rook-version=$TO_TAG/g" \
+     -e "s/--branch $FROM_TAG/--branch $TO_TAG/g" \
+  Documentation/Upgrade/rook-upgrade.md
+
+echo "Updating upgrade guide from $FROM_TAG_UPGRADE to $TO_TAG..."
+
+$SED -e "s/rook-version=$FROM_TAG_UPGRADE/rook-version=$TO_TAG/g" \
+    -e "s/rook\/ceph:$FROM_TAG_UPGRADE/rook\/ceph:$TO_TAG/g" \
+    -e "s/\`$FROM_TAG_UPGRADE\`/\`$TO_TAG\`/g" \
+    -e "s/when Rook $FROM_TAG_UPGRADE/when Rook $TO_TAG/g" \
+  Documentation/Upgrade/rook-upgrade.md
+
+echo "Done! Now you can open a PR with these changes."
```
There is a corner case here when we are branching new release docs from master. In that case, we are updating both current/new version and previous/deprecated versions. Do you want to handle that case with this script? IMO, it might be good to defer this for later, but I think it might be something to at least note in a comment so we can be reminded that the script is really only fully featured for z-stream updates but not new branch upgrades.

Screenshot b/c formatting messes up when copying code with code blocks:

<img width="781" alt="image" src="https://github.com/user-attachments/assets/f0bb7969-6e07-41ec-943f-24ac349a22e8" />

If this were the master branch and we were starting the v1.18 docs, there would be 2 updates here
1. Change v1.17.0 to either v1.18.0
2. Change v1.16.* to v1.17.Z (where Z is the latest 1.17)

BlaineEXE replied:
Sounds like the right approach to me. Thanks

BlaineEXE reviewed (approved):
This looks good to merge to me, with a note/question about a corner case and questions about if we want to handle that here: https://github.com/rook/rook/pull/15781/files#r2070521987
